packages/test-sync/src/commonMain/kotlin/io/realm/kotlin/test/mongodb/util/SchemaProcessor.kt

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2024 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+@file:Suppress("invisible_member", "invisible_reference")
+
+package io.realm.kotlin.test.mongodb.util
+
+import io.realm.kotlin.internal.interop.CollectionType
+import io.realm.kotlin.internal.interop.PropertyInfo
+import io.realm.kotlin.internal.interop.PropertyType
+import io.realm.kotlin.internal.realmObjectCompanionOrNull
+import io.realm.kotlin.internal.schema.RealmClassImpl
+import io.realm.kotlin.schema.RealmClassKind
+import io.realm.kotlin.types.BaseRealmObject
+import kotlin.reflect.KClass
+
+// TODO REname methods and classes
+class SchemaProcessor private constructor(
+    classes: Set<KClass<out BaseRealmObject>>,
+    private val databaseName: String,
+    private val extraProperties: Map<String, PrimitivePropertyType.Type>,
+) {
+    companion object {
+
+        fun process(
+            databaseName: String,
+            classes: Set<KClass<out BaseRealmObject>>,
+            extraProperties: Map<String, PrimitivePropertyType.Type> = emptyMap(),
+        ): Map<String, Schema> {
+            val processor = SchemaProcessor(classes, databaseName, extraProperties)
+
+            return processor.processedSchemas
+                .entries
+                .filterNot { (_, schema) -> schema.kind == RealmClassKind.EMBEDDED }
+                .associate { (name, schema) ->
+                    // add metadata
+                    name to Schema(
+                        databaseName,
+                        schema,
+                        processor.processedRelationships[name]!!
+                    )
+                }
+        }
+    }
+
+    private val realmSchemas: Map<String, RealmClassImpl> = classes.associate { clazz ->
+        val companion = clazz.realmObjectCompanionOrNull()!!
+        val realmSchema = companion.io_realm_kotlin_schema()
+        realmSchema.cinteropClass.name to realmSchema
+    }
+
+    val processedSchemas: MutableMap<String, SchemaData> = mutableMapOf()
+    val processedRelationships: MutableMap<String, Map<String, SchemaRelationship>> = mutableMapOf()
+
+    init {
+        // TODO CHECK embedded CYCLES
```

Comment:
What happens with cycles right now? If not supported could we at least easily break/throw, so that we don't have dig in if it happens.

Reply:
I forgot to add it...